watcher/watcher.go

```diff
@@ -1513,94 +1520,128 @@ func (w *Watcher) checkIfUsageIsAboveLimit(
-			}
+		return nil
+	}
 
-			if event.Type == watch.Error {
-				return nil
-			}
+	// Update pod in redis when it exists in informer cache
 
-			switch event.Type {
-			case watch.Added, watch.Modified:
-				logger := w.Logger.WithFields(logrus.Fields{
-					"operation": "watcher.kubeWatch.CreateOrUpdatePod",
-				})
+	logger := w.Logger.WithFields(logrus.Fields{
+		"operation": "watcher.kubeWatch.CreateOrUpdatePod",
+	})
 
-				// logger.Debug("new pod detected: ", event.Type)
-				if kubePod, ok := event.Object.(*v1.Pod); ok {
-					// create Pod from v1.Pod
-					pod := &models.Pod{
-						Name:          kubePod.GetName(),
-						Version:       kubePod.GetLabels()["version"],
-						NodeName:      kubePod.Spec.NodeName,
-						Status:        kubePod.Status,
-						Spec:          kubePod.Spec,
-						IsTerminating: models.IsPodTerminating(kubePod),
-					}
-
-					err := models.AddToPodMap(w.RedisClient.Client, w.MetricsReporter, pod, w.SchedulerName)
-					if err != nil {
-						logger.WithError(err).Error("failed to add pod to redis podMap key")
-					}
-				} else {
-					logger.Error("obj received is not of type *v1.Pod")
-				}
-			case watch.Deleted:
-				logger := w.Logger.WithFields(logrus.Fields{
-					"operation": "watcher.kubeWatch.DeletePod",
-				})
+	pod := &models.Pod{
+		Name:          kubePod.GetName(),
+		Version:       kubePod.GetLabels()["version"],
+		NodeName:      kubePod.Spec.NodeName,
+		Status:        kubePod.Status,
+		Spec:          kubePod.Spec,
+		IsTerminating: models.IsPodTerminating(kubePod),
+	}
 
-				// logger.Debug("new pod removed")
-				if kubePod, ok := event.Object.(*v1.Pod); ok {
-					// Remove pod from redis
-					err := models.RemoveFromPodMap(w.RedisClient.Client, w.MetricsReporter, kubePod.GetName(), w.SchedulerName)
-					if err != nil {
-						logger.WithError(err).Errorf("failed to remove pod %s from redis", kubePod.GetName())
-					}
-					room := models.NewRoom(kubePod.GetName(), w.SchedulerName)
-					err = room.ClearAll(w.RedisClient.Client, w.MetricsReporter)
-					if err != nil {
-						logger.WithError(err).Errorf("failed to clearAll %s from redis", kubePod.GetName())
-					}
-				} else {
-					logger.Error("obj received is not of type *v1.Pod or cache.DeletedFinalStateUnknown")
-				}
-			}
-		}
+	err = models.AddToPodMap(w.RedisClient.Client, w.MetricsReporter, pod, w.SchedulerName)
+	if err != nil {
+		logger.WithError(err).Error("failed to add pod to redis podMap key")
+		return err
 	}
+
 	return nil
 }
 
-func (w *Watcher) configureKubeWatch(stopCh <-chan struct{}) error {
+func (w *Watcher) processItem(obj interface{}) {
+	// Always need to mark obj as done in queue
+	defer w.Queue.Done(obj)
+
+	key, ok := obj.(string)
+	if !ok {
+		// Object is not a string so we skip, shouldn't happen
+		w.Queue.Forget(obj)
+		return
+	}
+
+	err := w.podEventHandler(key)
+	if err != nil {
+		// An error happened so we requeue it so it can processed again later
+		w.Queue.AddRateLimited(key)
```

Comment:
Do we want to reprocess in all error cases?

Reply:
Even though I treat some different errors in handler code the only errors that can really happen are redis errors and I think we always want to retry in this cases or redis state can become too different from Kubernetes and that's really bad for maestro.

Reply:
Cool, makes sense.